src/canopy/chat_engine/query_generator/condensed.py

```diff
@@ -0,0 +1,139 @@
+import logging
+import re
+from typing import List, Optional, cast
+
+from tenacity import retry, stop_after_attempt, retry_if_exception_type
+
+from canopy.chat_engine.models import HistoryPruningMethod
+from canopy.chat_engine.prompt_builder import PromptBuilder
+from canopy.chat_engine.query_generator import QueryGenerator, LastMessageQueryGenerator
+from canopy.llm import BaseLLM, AnyscaleLLM
+from canopy.models.data_models import Messages, Query, UserMessage
+from canopy.models.api_models import ChatResponse
+
+logger = logging.getLogger(__name__)
+
+SYSTEM_PROMPT = """
+You are an expert on formulating a search query for a search engine, to assist
+ in responding to the user's question.
+
+Given the following conversation, create a standalone question summarizing
+ the user's last question, in its original language.
+
+Reply to me in JSON in this format:
+
+ {"question": {The question you generated here}}.
+
+   Example:
+
+        User: What is the weather today?
+
+        Expected Response:
+            ```json
+            {"question": "What is the weather today?"}
+            ```
+
+    Example 2:
+
+        User: How should I wash my white clothes in the laundry?
+        Assistant: Separate from the colorful ones, and use a bleach.
+        User: Which temperature?
+
+        Expected Response:
+         ```json
+         {"question": "What is the right temperature for washing white clothes?"}
+         ```
+
+
+   Do not try to answer the question; just try to formulate a question representing the user's question.
+   Do not return any other format other than the specified JSON format and keep it really short.
+
+"""  # noqa: E501
+
+USER_PROMPT = "Return only a JSON that has 'question' as a key and the value."
+
+
+class ExtractionException(ValueError):
+    pass
+
+
+class CondensedQueryGenerator(QueryGenerator):
+    _DEFAULT_COMPONENTS = {
+        "llm": AnyscaleLLM,
+    }
+
+    def __init__(self,
+                 *,
+                 llm: Optional[BaseLLM] = None):
+        """
+             Tries to build a question out of all the chat history.
+             In order to do that, the system first tries to make LLM return
+             a JSON object with a key "question" and the value representing
+             the question the user was intending to ask. If LLM response
+             cannot be parsed it falls back to last message query generator,
+             meaning it returns the last message of the history as a query.
+        """
+        self._llm = llm or self._DEFAULT_COMPONENTS["llm"]()
+        self._system_prompt = SYSTEM_PROMPT
+        self._prompt_builder = PromptBuilder(HistoryPruningMethod.RAISE, 2)
+        self._last_message_query_generator = LastMessageQueryGenerator()
```

Comment:
Why compilation in this case and not inheritance?  

In this particular case, I think that inheriting `LastMessageQueryGenerator()` class, then in the fallback case calling `super().generate()` is the correct approach, isn't it?

Reply:
@igiloh-pinecone the inputs to the init function are completely different and I don't see any connection between them. It does not extend the functionality but more like calling the other in case it cannot generate. So I would want to keep it this way. Lmk what you think

Reply:
We'll agree to disagree 😄 .

Sure, no problem. Keep it this way.